Code/Engine/GameEngine/Gameplay/Implementation/RaycastComponent.cpp

```diff
@@ -0,0 +1,190 @@
+    EZ_MEMBER_PROPERTY("CollisionLayerTrigger", m_uiCollisionLayerTrigger)->AddAttributes(new ezDynamicEnumAttribute("PhysicsCollisionLayer")),
+    EZ_ACCESSOR_PROPERTY("TriggerMessage", GetTriggerMessage, SetTriggerMessage)
+  }
+  EZ_END_PROPERTIES;
+  EZ_BEGIN_ATTRIBUTES
+  {
+    new ezCategoryAttribute("Gameplay"),
+    new ezDirectionVisualizerAttribute(ezBasisAxis::PositiveX, 0.2f, ezColor::YellowGreen),
+  }
+  EZ_END_ATTRIBUTES;
+}
+EZ_END_DYNAMIC_REFLECTED_TYPE;
+// clang-format on
+
+ezRaycastComponent::ezRaycastComponent() = default;
+
+ezRaycastComponent::~ezRaycastComponent() = default;
+
+
+void ezRaycastComponent::OnSimulationStarted()
+{
+  m_pPhysicsWorldModule = GetWorld()->GetModule<ezPhysicsWorldModuleInterface>();
+  m_hLastTriggerObjectInRay.Invalidate();
+}
+
+void ezRaycastComponent::SerializeComponent(ezWorldWriter& stream) const
+{
+  SUPER::SerializeComponent(stream);
+  auto& s = stream.GetStream();
+
+  stream.WriteGameObjectHandle(m_hRaycastEndObject);
+  s << m_fMaxDistance;
+  s << m_bDisableTargetObjectOnNoHit;
+  s << m_uiCollisionLayerEndPoint;
+  s << m_uiCollisionLayerTrigger;
+  s << m_sTriggerMessage;
+}
+
+void ezRaycastComponent::DeserializeComponent(ezWorldReader& stream)
+{
+  SUPER::DeserializeComponent(stream);
+  auto& s = stream.GetStream();
+
+  m_hRaycastEndObject = stream.ReadGameObjectHandle();
+  s >> m_fMaxDistance;
+  s >> m_bDisableTargetObjectOnNoHit;
+  s >> m_uiCollisionLayerEndPoint;
+  s >> m_uiCollisionLayerTrigger;
+  s >> m_sTriggerMessage;
+}
+
+
+void ezRaycastComponent::SetTriggerMessage(const char* sz)
+{
+  m_sTriggerMessage.Assign(sz);
+}
+
+const char* ezRaycastComponent::GetTriggerMessage() const
+{
+  return m_sTriggerMessage.GetData();
+}
+
+void ezRaycastComponent::SetRaycastEndObject(const char* szReference)
+{
+  auto resolver = GetWorld()->GetGameObjectReferenceResolver();
+
+  if (!resolver.IsValid())
+    return;
+
+  m_hRaycastEndObject = resolver(szReference, GetHandle(), "RaycastEndObject");
+}
+
+void ezRaycastComponent::Update()
+{
+  if (m_hRaycastEndObject.IsInvalidated())
+    return;
+
+  ezGameObject* pEndObject = nullptr;
+  if (!GetWorld()->TryGetObject(m_hRaycastEndObject, pEndObject))
+  {
+    return;
+  }
+
+  if (!m_pPhysicsWorldModule)
+  {
+    // Happens in Prefab viewports
+    return;
+  }
+
+  const ezVec3 rayStartPosition = GetOwner()->GetGlobalPosition();
+  const ezVec3 rayDir = GetOwner()->GetGlobalDirForwards();
+
+  //ezDebugRenderer::Line lines[] = { {rayStartPosition, rayStartPosition + rayDir * m_fMaxDistance} };
+  //ezDebugRenderer::DrawLines(GetWorld(), lines, ezColor::RebeccaPurple);
+
+  float fHitDistance = m_fMaxDistance;
+  ezPhysicsHitResult hit;
+  if (m_pPhysicsWorldModule->CastRay(rayStartPosition, rayDir, m_fMaxDistance, m_uiCollisionLayerEndPoint, hit))
+  {
+    fHitDistance = hit.m_fDistance;
+
+    if (!pEndObject->IsActive() && m_bDisableTargetObjectOnNoHit)
+    {
+      pEndObject->SetActive(true);
```

Comment:
Just ran into this as well: We may need to enhance the active state on objects to be recursive and editable in the editor (it glitches when you expose it atm).